### ISSUE TITLE
Test failure regression in TestAzureVMResourceDetector

### ISSUE THREAD

**Issue by musicinmybrain** [open-telemetry/opentelemetry-python-contrib#2102]
**Describe your environment**

Fedora Linux 39, `x86_64`, Python 3.12.0, current `main` (772063ca8a4ac47b45301837e5b3f6f0a8edbe0a).

**Steps to reproduce**
Describe exactly how to reproduce the error. Include a code sample if applicable.

```
gh repo clone open-telemetry/opentelemetry-python-contrib
cd opentelemetry-python-contrib
cd resource/opentelemetry-resource-detector-azure/
python3 -m venv _e
. _e/bin/activate
pip install -e .[test]
pip install pytest
pytest
```

**What is the expected behavior?**

All tests pass.

**What is the actual behavior?**

```
================================================ test session starts =================================================
platform linux -- Python 3.12.0, pytest-7.4.3, pluggy-1.3.0 -- /home/ben/src/forks/opentelemetry-python-contrib/resource/opentelemetry-resource-detector-azure/_e/bin/python3
cachedir: .pytest_cache
rootdir: /home/ben/src/forks/opentelemetry-python-contrib
configfile: pytest.ini
collected 6 items                                                                                                    

tests/test_app_service.py::TestAzureAppServiceResourceDetector::test_off_app_service PASSED                    [ 16%]
tests/test_app_service.py::TestAzureAppServiceResourceDetector::test_on_app_service PASSED                     [ 33%]
tests/test_app_service.py::TestAzureAppServiceResourceDetector::test_on_app_service_no_owner PASSED            [ 50%]
tests/test_app_service.py::TestAzureAppServiceResourceDetector::test_on_app_service_no_resource_group PASSED   [ 66%]
tests/test_vm.py::TestAzureVMResourceDetector::test_linux 
--------------------------------------------------- live log call ----------------------------------------------------
ERROR    opentelemetry.resource.detector.azure.vm:vm.py:78 Failed to receive Azure VM metadata: 'str' object does not support the context manager protocol
Traceback (most recent call last):
  File "/home/ben/src/forks/opentelemetry-python-contrib/resource/opentelemetry-resource-detector-azure/src/opentelemetry/resource/detector/azure/vm.py", line 72, in get_azure_vm_metadata
    with urlopen(request).read() as response:
TypeError: 'str' object does not support the context manager protocol
FAILED                                                                                                         [ 83%]
tests/test_vm.py::TestAzureVMResourceDetector::test_windows 
--------------------------------------------------- live log call ----------------------------------------------------
ERROR    opentelemetry.resource.detector.azure.vm:vm.py:78 Failed to receive Azure VM metadata: 'str' object does not support the context manager protocol
Traceback (most recent call last):
  File "/home/ben/src/forks/opentelemetry-python-contrib/resource/opentelemetry-resource-detector-azure/src/opentelemetry/resource/detector/azure/vm.py", line 72, in get_azure_vm_metadata
    with urlopen(request).read() as response:
TypeError: 'str' object does not support the context manager protocol
FAILED                                                                                                         [100%]

====================================================== FAILURES ======================================================
_______________________________________ TestAzureVMResourceDetector.test_linux _______________________________________

self = <tests.test_vm.TestAzureVMResourceDetector testMethod=test_linux>
mock_urlopen = <MagicMock name='urlopen' id='139717256146928'>

    @patch("opentelemetry.resource.detector.azure.vm.urlopen")
    def test_linux(self, mock_urlopen):
        mock_open = Mock()
        mock_urlopen.return_value = mock_open
        mock_open.read.return_value = LINUX_JSON
        attributes = AzureVMResourceDetector().detect().attributes
        for attribute_key, attribute_value in LINUX_ATTRIBUTES.items():
>           self.assertEqual(attributes[attribute_key], attribute_value)

tests/test_vm.py:371: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = BoundedAttributes({}, maxlen=None), key = 'azure.vm.scaleset.name'

    def __getitem__(self, key):
>       return self._dict[key]
E       KeyError: 'azure.vm.scaleset.name'

_e/lib64/python3.12/site-packages/opentelemetry/attributes/__init__.py:163: KeyError
------------------------------------------------- Captured log call --------------------------------------------------
ERROR    opentelemetry.resource.detector.azure.vm:vm.py:78 Failed to receive Azure VM metadata: 'str' object does not support the context manager protocol
Traceback (most recent call last):
  File "/home/ben/src/forks/opentelemetry-python-contrib/resource/opentelemetry-resource-detector-azure/src/opentelemetry/resource/detector/azure/vm.py", line 72, in get_azure_vm_metadata
    with urlopen(request).read() as response:
TypeError: 'str' object does not support the context manager protocol
______________________________________ TestAzureVMResourceDetector.test_windows ______________________________________

self = <tests.test_vm.TestAzureVMResourceDetector testMethod=test_windows>
mock_urlopen = <MagicMock name='urlopen' id='139717253074672'>

    @patch("opentelemetry.resource.detector.azure.vm.urlopen")
    def test_windows(self, mock_urlopen):
        mock_open = Mock()
        mock_urlopen.return_value = mock_open
        mock_open.read.return_value = WINDOWS_JSON
        attributes = AzureVMResourceDetector().detect().attributes
        for attribute_key, attribute_value in LINUX_ATTRIBUTES.items():
>           self.assertEqual(attributes[attribute_key], attribute_value)

tests/test_vm.py:380: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = BoundedAttributes({}, maxlen=None), key = 'azure.vm.scaleset.name'

    def __getitem__(self, key):
>       return self._dict[key]
E       KeyError: 'azure.vm.scaleset.name'

_e/lib64/python3.12/site-packages/opentelemetry/attributes/__init__.py:163: KeyError
------------------------------------------------- Captured log call --------------------------------------------------
ERROR    opentelemetry.resource.detector.azure.vm:vm.py:78 Failed to receive Azure VM metadata: 'str' object does not support the context manager protocol
Traceback (most recent call last):
  File "/home/ben/src/forks/opentelemetry-python-contrib/resource/opentelemetry-resource-detector-azure/src/opentelemetry/resource/detector/azure/vm.py", line 72, in get_azure_vm_metadata
    with urlopen(request).read() as response:
TypeError: 'str' object does not support the context manager protocol
============================================ 2 failed, 4 passed in 0.09s =============================================
```

**Additional context**

Originally observed while updating Fedora Linux Rawhide to the 1.22.0/0.43b0 release (`v0.43b0`); I can also reproduce it using that tag. This appears to be a regression from `v0.42b0`, where all tests pass.

**Comment by musicinmybrain** [comment]
I know this project is still working on officially supporting Python 3.12, so I tried again with Python 3.11.6 and was still able to reproduce the failures.

**Comment by musicinmybrain** [comment]
This is still present in https://pypi.org/project/opentelemetry-resource-detector-azure/0.1.3/.

**Comment by xrmx** [comment]
Looks like this has been fixed, at least main tests run fine here. @musicinmybrain any chance you can check please?

**Comment by musicinmybrain** [comment]
> Looks like this has been fixed, at least main tests run fine here. @musicinmybrain any chance you can check please?

I repeated this:

> ```
> gh repo clone open-telemetry/opentelemetry-python-contrib
> cd opentelemetry-python-contrib
> cd resource/opentelemetry-resource-detector-azure/
> python3 -m venv _e
> . _e/bin/activate
> pip install -e .[test]
> pip install pytest
> pytest
> ```

I had to add `pip install opentelemetry-instrumentation` because the `test` extras were removed since I originally reported this, but then all tests passed.

Removing the test skips in the [Fedora package](https://src.fedoraproject.org/rpms/python-opentelemetry-resource-detector-azure) also works.

Thus, I would say this is fixed in 0.1.5 (or earlier).

Thanks!
